### ISSUE TITLE
source shape as geometric object

### ISSUE THREAD

**Issue by oskooi** [NanoComp/meep#612]
Currently, the shape of a [`Source`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#source) can only be a rectangular prism defined by its `size` property. Although the amplitude function property (`amp_func`) does enable arbitrary shapes, this is cumbersome to set up for simple geometries. It would be useful if the source shape could also be defined by a [`GeometricObject`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#geometricobject) (e.g., `Sphere`, `Cylinder`, `Prism`, etc.).

**Comment by oskooi** [comment]
One application would be for designing grating in-couplers for silicon photonics where a [cleaved fiber](https://en.wikipedia.org/wiki/Cleave_(fiber)) is used as the input source and placed in the near-field of the grating. See e.g., [Optics Express, Vol. 18, pp. 18278-83, 2010](https://www.osapublishing.org/oe/abstract.cfm?uri=oe-18-17-18278). Another application would be to model a [VCSEL](https://en.wikipedia.org/wiki/Vertical-cavity_surface-emitting_laser) source, which also tend to have cylindrical cross sections, for use in e.g., next-generation displays such as AR/MR smart glasses.

Once this feature is ready, we will then need to modify the [eigenmode source](https://meep.readthedocs.io/en/latest/Python_User_Interface/#eigenmodesource) such that even though it is computed from a given volume/cross-section it can be placed anywhere in the cell.

**Comment by stevengj** [comment]
@oskooi, I don't think we actually need a geometric-object source for that application:

* As you say, the eigenmode source would have to be modified anyway to map it to an oblique slice, and once you did that you probably wouldn't need the geometric-object source.

* Even easier, you can take a Cartesian x/y/z slice of an oblique waveguide and use an *ordinary* eigenmode source, with one modification — tell MPB that the third lattice vector is in the oblique direction, not orthogonal to the plane.   This should make it compute an oblique cross-section of the mode traveling down your oblique waveguide, and create the corresponding current source via the principle of equivalence.

**Comment by oskooi** [comment]
The following schematic demonstrates the functionality that we are trying to enable. An eigenmode source is defined for a Cartesian *x*/*y*/*z* slice of an oblique waveguide which generates a unidirectional propagating mode. Telling MPB that the mode's wavevector is not orthogonal to the slice is not enough; we would also like to be able to specify its `eig_parity` somehow just as we would if the waveguide is oriented along the *x* axis and we could specify either `EVEN_Y` or `ODD_Y`.

![eigenmode_source_oblique](https://user-images.githubusercontent.com/7152530/51090804-c5759100-1736-11e9-9fd1-5ce0c131d3f4.png)

**Comment by stevengj** [comment]
It looks like #675 addresses the oblique-waveguide source motivation of this issue.

However, we might as well leave this open.   There are occasionally circumstances where you want a finite-volume source instead of a point source, and it would be nice to support a spherical source volume etcetera with subpixel averaging.   Implementing subpixel averaging of the edges is something that requires a low-level C++ implementation, where it is straightforward as I [commented in #635](https://github.com/NanoComp/meep/pull/635#issuecomment-450660248), and cannot easily be done in user-level Python code.

But it doesn't seem like a high-priority feature.
